packages/widgets/src/menubar.ts

```diff
@@ -479,9 +460,72 @@ export class MenuBar extends Widget {
           this.activeIndex = i;
         }
       });
+      // Calculate size of current menu
+      totalMenuSize += this._menuItemSizes[i];
+    }
+    // Render overflow menu if needed
+    if (this._overflowIndex > -1) {
+      // Create overflow menu
+      if (this._overflowMenu === null) {
+        this._overflowMenu = new Menu({ commands: new CommandRegistry() });
+        this._overflowMenu.title.label = '...';
+        this._overflowMenu.title.mnemonic = 0;
+        this.addMenu(this._overflowMenu, false);
+      }
+      // Move menus to overflow menu
+      for (let i = length; i < menus.length - 1; ++i) {
+        let submenu = this.menus[i];
+        submenu.title.mnemonic = 0;
+        this._overflowMenu.insertItem(0, {
+          type: 'submenu',
+          submenu: submenu
+        });
+        this.removeMenu(submenu, false);
+      }
+      let title = this._overflowMenu.title;
+      let active = length === activeIndex;
+      if (active && menus[length].items.length == 0) {
+        active = false;
+      }
+      content[length] = renderer.renderItem({
+        title,
+        active,
+        onfocus: () => {
+          this.activeIndex = length;
+        }
+      });
+    } else if (this._overflowMenu !== null) {
+      // Remove submenus from overflow menu
+      let overflowMenuItems = this._overflowMenu.items;
+      let screenSize = this.node.offsetWidth;
+      let n = this._overflowMenu.items.length;
+      for (let i = 0; i < n; ++i) {
+        let index = menus.length - 1 - i;
+        if (screenSize - totalMenuSize > this._menuItemSizes[index]) {
+          let menu = overflowMenuItems[0].submenu as Menu;
+          this._overflowMenu.removeItemAt(0);
+          this.insertMenu(length, menu, false);
+          let title = menu.title;
+          let active = false;
+          content[length] = renderer.renderItem({
+            title,
+            active,
+            onfocus: () => {
+              this.activeIndex = length;
+            }
+          });
+          length++;
+        }
+      }
+      if (this._overflowMenu.items.length === 0) {
+        this.removeMenu(this._overflowMenu, false);
+        content.pop();
+        this._overflowMenu = null;
+        this._overflowIndex = -1;
+      }
     }
-    this._menuItemSizes = [];
     VirtualDOM.render(content, this.contentNode);
+    this.updateOverflowIndex();
```

Comment:
It looks like this one will lead to the state being always delayed by one tick. It is not bad in itself if this is a conscious decision made as a performance-UX tradeoff. Could we document this?

One thing to consider is whether we should wrap it in an extra `requestAnimationFrame` to protect against layout trashing.

Reply:
How do I wrap it in a `requestAnimationFrame`?

Reply:
It would be something like:

```js
requestAnimationFrame(() => this._updateOverflowIndex());
```

I have not tested if this helps. You would need to record a profile in developer tools when you resize the page to trigger collapsing before and after applying the change and compare. If you have the profiles it should be obvious from the time spend on forced style recalculation/layout whether it is worth waiting for the next frame or not.